flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -1515,6 +1519,62 @@ mlir::Value IntrinsicLibrary::genAbs(mlir::Type resultType,
   llvm_unreachable("unexpected type in ABS argument");
 }
 
+// ADJUSTL
+fir::ExtendedValue
+IntrinsicLibrary::genAdjustL(mlir::Type resultType,
```

Comment:
You could share `genTrim`, `genAdjustL`, genAdjustR` implementations that are very similar into a single function to centralize maintenance.

Adding a declaration in the IntrinsicLibrary class like the following:
```
  template <void (*CallRuntime)(Fortran::lower::FirOpBuilder &,
                                mlir::Location loc, mlir::Value, mlir::Value)>
     fir::ExtendedValue genCharDescriptorRtCall(mlir::Type,
                                   llvm::ArrayRef<fir::ExtendedValue>);
```

Replacing genTrim implementation and similar by:
```
template<void (*CallRuntime)(Fortran::lower::FirOpBuilder &, mlir::Location loc, mlir::Value, mlir::Value)>
fir::ExtendedValue IntrinsicLibrary::genCharDescriptorRtCall(mlir::Type resultType,
                          llvm::ArrayRef<fir::ExtendedValue> args) {
  // ...
  // Call the runtime -- the runtime will allocate the result.
  CallRuntime(builder, loc, resultIrBox, string);
  // ...
}
```

And in the table:

```
{"adjustl", &I::genCharDescriptorRtCall<Fortran::lower::genAdjustL>,
   {{{"string", asAddr}}},
    /*isElemental=*/true},
```

It's not a big deal if the `fir::emitFatalError` does not give the intrinsic name, with the location, there is already enough information for this unlikely internal error, and you are more than welcome if you have a better name than `genCharDescriptorRtCall`.

Now, the duplicated logic is small and rather simple, so if you think this is overkill and that sticking with the pattern is better, that may also be OK.

Reply:
I pondered this briefly but didn't dig into all the possible intrinsics that might match this particular case.  Given the relative simplicity here, I decided to skip it as a first cut.  I'll revisit it -- my first inclination is to share between the two `adjust` calls but skip rolling `trim` into the mix.  The primary reasoning being to provide some clarity in naming and navigating the patterns across lowering the intrinsics.  

In the end, I don't think it is big deal either way.  Happy to push up a change along either route -- thoughts?

Reply:
I am fine with a solution sharing only adjutl and adjustr. I agree it is easier to reason about.

Reply:
@jeanPerier -- pushed your suggested refactoring.  Take a look.  Thanks.